lib/money_column/active_record_hooks.rb

```diff
@@ -28,58 +28,78 @@ def init_internals
-            @money_column_cache.clear
-            super(value)
-          end
+          raise ArgumentError, 'must set one of :currency_column or :currency options'
         end
 
         columns.flatten.each do |column|
-          if currency_read_only || currency
-            money_column_reader(column, currency_column, currency_object)
-            money_column_writer(column, currency_column, currency_object)
-          else
-            composed_of(
-              column.to_sym,
-              class_name: 'Money',
-              mapping: [[column.to_s, 'value'], [currency_column.to_s, 'currency']],
-              converter: Proc.new { |value| value.to_money },
-              allow_nil: true,
-            )
-          end
+          money_column_reader(column, currency_column, currency_iso)
+          money_column_writer(column, currency_column, currency_iso, currency_read_only)
+          money_column_reflection(column, currency_column, currency_read_only)
         end
       end
 
       private
 
-      def money_column_reader(column, currency_column, currency_object)
+      def money_column_reflection(column, currency_column, currency_read_only)
+        options = {
+          class_name: 'Money',
+          mapping: currency_read_only ? [column.to_s, 'value'] : [[column.to_s, 'value'], [currency_column.to_s, 'currency_iso']],
+          allow_nil: true
+        }
+        reflection = ActiveRecord::Reflection.create(:composed_of, column, nil, options, self)
+        ActiveRecord::Reflection.add_aggregate_reflection self, column, reflection
+      end
+
+      def clear_cache_on_currency_change(currency_column)
+        return unless currency_column
+        define_method "#{currency_column}=" do |value|
+          @money_column_cache.clear
+          super(value)
+        end
+      end
+
+      def money_column_reader(column, currency_column, currency_iso)
         define_method column do
           return @money_column_cache[column] if @money_column_cache[column]
           return unless value = read_attribute(column)
-          iso = currency_object || try(currency_column)
+          iso = currency_iso || try(currency_column)
           @money_column_cache[column] = Money.new(value, iso)
         end
       end
 
-      def money_column_writer(column, currency_column, currency_object)
+      def money_column_writer(column, currency_column, currency_iso, currency_read_only)
         define_method "#{column}=" do |money|
           if money.blank?
             write_attribute(column, nil)
             return @money_column_cache[column] = nil
           end
-          money = money.to_money
-          currency_db = Money::Helpers.value_to_currency(currency_object || try(currency_column))
 
-          unless currency_db && currency_db.compatible?(money.currency)
-            Money.deprecate("[money_column] currency mismatch between #{currency_db} and #{money.currency}.")
+          currency_db = currency_iso || try(currency_column)
```

Comment:
Maybe call this `currency_source` because a statically defined currency code doesn't come from the DB.